### DISCUSSION TITLE
yamlfmt's recent slow development

### DISCUSSION THREAD

**Discussion by braydonk** [google/yamlfmt#149] [General]
Hello yamlfmt users!

Thank you very much for using the tool, and for all the feedback and contributions. I really hope that this tool has been useful to you.It has been exciting to see projects even adopting it as an underlying library, and it was so wild to use it accidentally when I set up `mason.nvim` for the first time.

I wanted to address the fact that while I have been responding on issues generally as soon as I see them (occasionally I do miss them but I try), my development work has been slow. My most recent feature was when I managed to get together some integration tests, which isn't even user-visible.

I understand that this slowed development can be frustrating for a tool that folks rely on. I do want to find a way to pick up the pace a little bit, but it has been challenging for reasons I'll get into. 

## TL;DR

* `yamlfmt`'s success on launch was a fluke I wasn't prepared for
* I never really adapted properly, change in work responsibilities restricted my ability to spend work time on it, and spare time was constrained in other ways
* I am trying to change things on the work-time front, and I will work to open up `yamlfmt` to easier external contribution, reach out if you are interested
* You are cool :slightly_smiling_face: 

## Project Origins

First, I should talk about the origins of this project and the accidental nature of it. I wrote the original super-minimal version of this tool during a tech-debt sprint on my team. The task was to add a yaml formatting tool to our repos so we could avoid scenarios where uncaught slight yaml mistakes were breaking parts of our automation. I thought this would be a quick task, but when I went looking for a simple formatting tool I came up shorthanded. Most of the options available required either Python or Node to be installed, something I thought seemed unnecessary and that I didn't want to force on my team just to format yaml. I thought "How hard could it be to just unmarshal the yaml and immediately marshal it to get a consistent output?" Turns out, the original version of this wasn't that hard. However, on a whim at the last second I figured I would take the extra time, make it into a separate tool that lived in its own repo, and add some minor configuration options. I decided to launch this repo under the `google` org, since it's a tool that we will be using on my team that I made on company time.

A day after the repo went live, it had over 200 stars on GitHub. I don't think I can overstate just how shocking this was. I figured it might get a bit more visibility being on the `google` GitHub org and being an intuitively google-able name, but I did not expect anything like that. It was exciting, but also a bit of a panic spiral which you can see in a [tweet thread I made after the uptick](https://twitter.com/RageCage64_/status/1561379869321269251). 

I was woefully unprepared to have so many people look at this project at once. I decided to do everything I could to get it ready. I was afforded an extra half-a-sprint to address the immediate concerns that I did not expect to need to address so soon, and decided to drop my existing [20% project](https://en.wikipedia.org/wiki/Side_project_time#Google_implementation) and make `yamlfmt` my 20% project instead. While using 20% time was good, the bulk of my work was done by putting in quite a bit of spare time outside of work.

## How I (try to) organize my time

I went into the history of this project to demonstrate that one of the reasons I've been having this trouble is because I did not make this project with the intention of it taking up, realistically, any of my time after launching. After the success of the project I have tried my best to allocate the time that the project deserves, but it has been a losing battle.

(I debated including this part in the post, but have opted to do so because it's the only way to realistically explain where the time went)  
Over the past 8 months or so, my ability to spend 20% time on this project has largely fell apart. I quickly became more busy with work responsibilities that took priority over my 20% project. Since the project's inception, this tool has largely not been recognized as part of my work responsibilities. Any of my work time spent on this was time not spent on things that were an actual priority for the team, and as a result I had to put that first. A lesser factor but still part of it is that I have been working to try and get a promotion, so I had to make more time for projects that had more of the sorts of required impact. (I'm aware that some of this is on me for not advocating for the project well enough internally; it's one of my weakpoints that I'm trying to address.) 

As a result, I am largely only able to spend spare time on this project. I try to carve out time to work on it when I can, as I am sad that there are some user issues on this repo that have been open as long as a year. I have also been watching the star count creep up over time, as well as the `Used By` column. Since creating this project as well, my involvement in other open source projects (primarily in the observability space) has greatly increased which also eats into my time.

## The Future

I am trying really hard to keep the project alive. After some recent developments at work, it looks like there is a possibility that I will be able to return to spending 20% time on this project (that isn't a guarantee, but I will be attempting to advocate for this).

My top priority over the next couple weeks is to properly beef up the Contributing Guidelines, PR/Issue Templates, and developer documentation to try and open up the floor for contributors. Even if I can get 20% time again, I don't think I will be able to drastically up the amount of time I am able to put towards the project. I will try to get this project more readily prepared for external contribution. If you are interested in contributing to this project, please get in touch. If enough people are interested I will try to set up a private discussion space.

## Thank you

Thank you for using the tool, and if you have been frustrated with a lack of response from me, or slow feature development, I hope that this post can address some of that even if it can't fix it.
